File: src/BrewUp.Rest.FSharp/Modules/OpenApiModule.fs
namespace BrewUp.Rest.FSharp.Modules

open Microsoft.AspNetCore.Builder
open Microsoft.Extensions.DependencyInjection
open Microsoft.OpenApi.Models

type OpenApiModule() =
    interface IModule with
        member _.IsEnabled = true
        member _.Order = 0
        
        member _.Register(builder: WebApplicationBuilder) =
            builder.Services.AddEndpointsApiExplorer() |> ignore
            builder.Services.AddSwaggerGen(fun c ->
                c.SwaggerDoc("v1", OpenApiInfo(Title = "BrewUp API", Version = "v1"))
            ) |> ignore
            builder.Services
            
        member _.Configure(app: WebApplication) =
            if app.Environment.EnvironmentName = "Development" then
                app.UseSwagger() |> ignore
                app.UseSwaggerUI() |> ignore
            app

File: src/BrewUp.Rest/Modules/OpenApiModule.cs
using Microsoft.OpenApi;
using Scalar.AspNetCore;

namespace BrewUp.Rest.Modules;

public static class OpenApiModule
{
  public static IServiceCollection Register(WebApplicationBuilder builder)
  {
    builder.Services.AddOpenApi(options =>
    {
      options.AddDocumentTransformer((document, _, _) =>
          {
            document.Servers = [new OpenApiServer { Url = "/" }];
            document.Info = new OpenApiInfo
            {
              Title = "BrewUp API",
              Version = "v1.0",
              Description = "BrewUp API for managing a Brewery",
              Contact = new OpenApiContact
              {
                Name = "BrewUp"
              }
            };

            return Task.CompletedTask;
          });
    });

    return builder.Services;
  }

  public static WebApplication Configure(WebApplication app)
  {
    app.MapOpenApi();
    app.MapScalarApiReference(options =>
    {
      options.WithTitle("BrewUp API")
              .WithTheme(ScalarTheme.None);
    });

    return app;
  }
}